libbpf-tools/biosnoop.bpf.c

```diff
@@ -0,0 +1,153 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright (c) 2020 Wenbo Zhang
+#include "vmlinux.h"
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_core_read.h>
+#include <bpf/bpf_tracing.h>
+#include "biosnoop.h"
+
+#define MAX_ENTRIES 10240
+
+const volatile char targ_disk[DISK_NAME_LEN] = {};
+const volatile bool targ_queued = false;
+
+struct piddata {
+	char comm[TASK_COMM_LEN];
+	u32 pid;
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, struct request *);
+	__type(value, struct piddata);
+	__uint(map_flags, BPF_F_NO_PREALLOC);
+} infobyreq SEC(".maps");
+
+struct stage {
+	u64 insert;
+	u64 issue;
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, struct request *);
+	__type(value, struct stage);
+} start SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERF_EVENT_ARRAY);
+	__uint(key_size, sizeof(u32));
+	__uint(value_size, sizeof(u32));
+} events SEC(".maps");
+
+SEC("fentry/blk_account_io_start")
+int BPF_PROG(tp_btf__blk_account_io_start, struct request *rq, bool new_io)
+{
+	u64 id = bpf_get_current_pid_tgid();
+	struct piddata piddata = {};
+
+	piddata.pid = id;
+	bpf_get_current_comm(&piddata.comm, sizeof(&piddata.comm));
+	bpf_map_update_elem(&infobyreq, &rq, &piddata, 0);
+	return 0;
+}
+
+static __always_inline bool disk_filtered(const char *disk)
+{
+	int i;
+
+	for (i = 0; targ_disk[i] != '\0' && i < DISK_NAME_LEN; i++) {
+		if (disk[i] != targ_disk[i])
+			return false;
+	}
+	return true;
+}
+
+static __always_inline
+int trace_rq_start(struct request *rq, bool insert)
+{
+	struct stage *stagep, stage = {};
+	u64 ts = bpf_ktime_get_ns();
+	char disk[DISK_NAME_LEN];
+
+	stagep = bpf_map_lookup_elem(&start, &rq);
+	if (!stagep) {
+		bpf_probe_read_kernel_str(&disk, sizeof(disk),
+					rq->rq_disk->disk_name);
+		if (!disk_filtered(disk)) {
+			bpf_map_delete_elem(&infobyreq, &rq);
+			return 0;
+		}
+		stagep = &stage;
+	}
+	if (insert)
+		stagep->insert = ts;
+	else
+		stagep->issue = ts;
+	bpf_map_update_elem(&start, &rq, stagep, BPF_NOEXIST);
```

Comment:
why even attempting to do bpf_map_update_elem with BPF_NOEXIST for case where we did find element in start? Just wasted work, no?

Or it's possible for some other CPU to insert element between lookup at line 75 and update at line 89?

Reply:
Yes, just wasted work. Fixed, thanks.